### PR TITLE
[SPARK-37893][CORE][TESTS] Avoid ConcurrentModificationException related to SparkFunSuite.LogAppender#_threshold"

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
When `mvn clean install -pl sql/core -am -Pscala-2.13` is executed, `AdaptiveQueryExecSuite` failed with a very small probability, the error stack as follows:

```
- Logging plan changes for AQE *** FAILED ***
  java.util.ConcurrentModificationException: mutation occurred during iteration
  at scala.collection.mutable.MutationTracker$.checkMutations(MutationTracker.scala:43)
  at scala.collection.mutable.CheckedIndexedSeqView$CheckedIterator.hasNext(CheckedIndexedSeqView.scala:47)
  at scala.collection.StrictOptimizedIterableOps.filterImpl(StrictOptimizedIterableOps.scala:225)
  at scala.collection.StrictOptimizedIterableOps.filterImpl$(StrictOptimizedIterableOps.scala:222)
  at scala.collection.mutable.ArrayBuffer.filterImpl(ArrayBuffer.scala:43)
  at scala.collection.StrictOptimizedIterableOps.filterNot(StrictOptimizedIterableOps.scala:220)
  at scala.collection.StrictOptimizedIterableOps.filterNot$(StrictOptimizedIterableOps.scala:220)
  at scala.collection.mutable.ArrayBuffer.filterNot(ArrayBuffer.scala:43)
  at org.apache.spark.SparkFunSuite$LogAppender.loggingEvents(SparkFunSuite.scala:288)
  at org.apache.spark.sql.execution.adaptive.AdaptiveQueryExecSuite.$anonfun$new$152(AdaptiveQueryExecSuite.scala:1487)
```

So this pr adds synchronous access control for `SparkFunSuite$LogAppender.loggingEvents` to avoid the above exceptions

### Why are the changes needed?
Bug fix

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GA

### CODE REVIEW

**Comment by LuciferYang** [comment]
The above UT failure is a very small probability event, cc @viirya

File: core/src/test/scala/org/apache/spark/SparkFunSuite.scala [hunk 1]
```diff
@@ -272,19 +272,23 @@ abstract class SparkFunSuite
...
     }
 
     def setThreshold(threshold: Level): Unit = {
       _threshold = threshold
     }
 
-    def loggingEvents: ArrayBuffer[LogEvent] = _loggingEvents.filterNot(_ == null)
+    def loggingEvents: ArrayBuffer[LogEvent] = _loggingEvents.synchronized{
+      _loggingEvents.filterNot(_ == null)
+    }
```
Comment (viirya) [inline core/src/test/scala/org/apache/spark/SparkFunSuite.scala:292]: `_loggingEvents.synchronized { ... }`

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
cc @gengliangwang

File: core/src/test/scala/org/apache/spark/SparkFunSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/test/scala/org/apache/spark/SparkFunSuite.scala:292]: +1 for @viirya 's comment.

**Review by dongjoon-hyun — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/SparkFunSuite.scala (hunk 1 above)
Comment (LuciferYang) [reply core/src/test/scala/org/apache/spark/SparkFunSuite.scala]: Thanks @dongjoon-hyun

**Review by gengliangwang — APPROVED** [review]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you all. Merged to master!

**Comment by dongjoon-hyun** [comment]
Uh, it looks wrong. Something goes wrong during merging via the merge script. I only changed one line here, but it causes the following.

```
Lead-authored-by: Dongjoon Hyun <dongjoon@apache.org>
Co-authored-by: yangjie01 <yangjie01@baidu.com>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>
```

**Comment by dongjoon-hyun** [comment]
Very sorry, @LuciferYang . The merge script didn't work like this so far. :(

**Comment by dongjoon-hyun** [comment]
In your branch, the ownership was clean line by line. Let me check the bug in our merge script.
```
7fd361973d2 (Liang-Chi Hsieh 2021-12-23 19:41:02 -0800 272)     override def append(loggingEvent: LogEvent): Unit = loggingEvent.synchronized {
7fd361973d2 (Liang-Chi Hsieh 2021-12-23 19:41:02 -0800 273)       val copyEvent = loggingEvent.toImmutable
7fd361973d2 (Liang-Chi Hsieh 2021-12-23 19:41:02 -0800 274)       if (copyEvent.getLevel.isMoreSpecificThan(_threshold)) {
0e3dcdbd7f1 (yangjie01       2022-01-13 17:22:53 +0800 275)         _loggingEvents.synchronized {
0e3dcdbd7f1 (yangjie01       2022-01-13 17:22:53 +0800 276)           if (_loggingEvents.size >= maxEvents) {
0e3dcdbd7f1 (yangjie01       2022-01-13 17:22:53 +0800 277)             val loggingInfo = if (msg == "") "." else s" while logging $msg."
0e3dcdbd7f1 (yangjie01       2022-01-13 17:22:53 +0800 278)             throw new IllegalStateException(
0e3dcdbd7f1 (yangjie01       2022-01-13 17:22:53 +0800 279)               s"Number of events reached the limit of $maxEvents$loggingInfo")
0e3dcdbd7f1 (yangjie01       2022-01-13 17:22:53 +0800 280)           }
0e3dcdbd7f1 (yangjie01       2022-01-13 17:22:53 +0800 281)           _loggingEvents.append(copyEvent)
ca5cabe5fbf (Liang-Chi Hsieh 2021-12-17 21:40:51 -0800 282)         }
88fc8dbc09c (Maxim Gekk      2020-01-14 16:03:10 +0900 283)       }
88fc8dbc09c (Maxim Gekk      2020-01-14 16:03:10 +0900 284)     }
```

**Comment by dongjoon-hyun** [comment]
Until now, I thought `Primary Author` is determined by the number of lines. It turns out that it's just a number of commits.
https://github.com/apache/spark/blob/1ef56e969f42726c630572f4e781cbeb3b57b888/dev/merge_spark_pr.py#L138-L146

More importantly, I didn't pay attention for the follow-up question. My bad. Sorry again, @LuciferYang . Next time, I'll not touch the PR in GitHub.

**Comment by dongjoon-hyun** [comment]
I'll make a PR to fix the bug.

**Comment by LuciferYang** [comment]
It doesn't matter ~ :)

**Pull request closed without merging**